dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java

```diff
@@ -17,1286 +17,404 @@
-                    fullName = defaultPath;
-                }
-                resourcesList =
-                        recursive ? storageOperate.listFilesStatusRecursively(fullName, defaultPath, tenantCode, type)
-                                : storageOperate.listFilesStatus(fullName, defaultPath, tenantCode, type);
-            } catch (Exception e) {
-                log.error(e.getMessage() + " Resource path: {}", fullName, e);
-                throw new ServiceException(String.format(e.getMessage() + " make sure resource path: %s exists in %s",
-                        defaultPath, resourceStorageType));
-            }
-        }
-
-        return resourcesList;
+        storageOperator.createStorageDir(createDirectoryDto.getDirectoryAbsolutePath());
+        log.info("Success create directory: {}", createDirectoryRequest.getParentAbsoluteDirectory());
     }
 
-    /**
-     * create directory
-     * xxx The steps to verify resources are cumbersome and can be optimized
-     *
-     * @param loginUser login user
-     * @param fullName  full name
-     * @param type      resource type
-     * @param result    Result
-     */
-    private void createDirectory(User loginUser, String fullName, ResourceType type, Result<Object> result) {
-        String tenantCode = tenantMapper.queryById(loginUser.getTenantId()).getTenantCode();
-        // String directoryName = storageOperate.getFileName(type, tenantCode, fullName);
-        String resourceRootPath = storageOperate.getDir(type, tenantCode);
+    @Override
+    public void createFile(CreateFileRequest createFileRequest) {
+        CreateFileDto createFileDto = createFileRequestTransformer.transform(createFileRequest);
+        createFileDtoValidator.validate(createFileDto);
+
+        // todo: use storage proxy
+        MultipartFile file = createFileDto.getFile();
+        String fileAbsolutePath = createFileDto.getFileAbsolutePath();
+        String srcLocalTmpFileAbsolutePath = copyFileToLocal(file);
         try {
-            if (!storageOperate.exists(resourceRootPath)) {
-                storageOperate.createTenantDirIfNotExists(tenantCode);
-            }
-
-            if (!storageOperate.mkdir(tenantCode, fullName)) {
-                throw new ServiceException(String.format("Create resource directory: %s failed.", fullName));
-            }
-            putMsg(result, Status.SUCCESS);
-        } catch (Exception e) {
-            throw new ServiceException(String.format("create resource directory: %s failed.", fullName));
+            storageOperator.upload(srcLocalTmpFileAbsolutePath, fileAbsolutePath, true, false);
+            ApiServerMetrics.recordApiResourceUploadSize(file.getSize());
+            log.info("Success upload resource file: {} complete.", fileAbsolutePath);
+        } catch (Exception ex) {
+            // If exception, clear the tmp path
+            FileUtils.deleteFile(srcLocalTmpFileAbsolutePath);
+            throw ex;
```

Comment:
It seems that the exception stack information cannot be seen here

Reply:
We directory the exception, so the exception stack will be log at `ApiExceptionHandler`